test/lit/passes/reorder-globals.wast

```diff
@@ -0,0 +1,376 @@
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT:  (drop
+  ;; RTRIP-NEXT:   (global.get $c)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT: )
+  (func $uses
+    (drop
+      (global.get $b)
+    )
+    (drop
+      (global.get $c)
+    )
+    (drop
+      (global.get $c)
+    )
+  )
+)
+
+;; As above, but a mixed case: $b depends on $a but $c has no dependencies. $c
+;; can be first.
+(module
+  ;; CHECK:      (global $c i32 (i32.const 30))
+
+  ;; CHECK:      (global $a i32 (i32.const 10))
+  ;; RTRIP:      (global $c i32 (i32.const 30))
+
+  ;; RTRIP:      (global $a i32 (i32.const 10))
+  (global $a i32 (i32.const 10))
+  ;; CHECK:      (global $b i32 (global.get $a))
+  ;; RTRIP:      (global $b i32 (global.get $a))
+  (global $b i32 (global.get $a))
+  (global $c i32 (i32.const 30))
+
+  ;; CHECK:      (func $uses
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (global.get $b)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (global.get $c)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (global.get $c)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  ;; RTRIP:      (func $uses
+  ;; RTRIP-NEXT:  (drop
+  ;; RTRIP-NEXT:   (global.get $b)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT:  (drop
+  ;; RTRIP-NEXT:   (global.get $c)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT:  (drop
+  ;; RTRIP-NEXT:   (global.get $c)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT: )
+  (func $uses
+    (drop
+      (global.get $b)
+    )
+    (drop
+      (global.get $c)
+    )
+    (drop
+      (global.get $c)
+    )
+  )
+)
+
+;; Another mixed case, now with $c depending on $b. $b can be before $a.
+(module
+  ;; CHECK:      (global $b i32 (i32.const 20))
+
+  ;; CHECK:      (global $c i32 (global.get $b))
+
+  ;; CHECK:      (global $a i32 (i32.const 10))
+  ;; RTRIP:      (global $b i32 (i32.const 20))
+
+  ;; RTRIP:      (global $c i32 (global.get $b))
+
+  ;; RTRIP:      (global $a i32 (i32.const 10))
+  (global $a i32 (i32.const 10))
+  (global $b i32 (i32.const 20))
+  (global $c i32 (global.get $b))
+
+  ;; CHECK:      (func $uses
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (global.get $b)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (global.get $c)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (global.get $c)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  ;; RTRIP:      (func $uses
+  ;; RTRIP-NEXT:  (drop
+  ;; RTRIP-NEXT:   (global.get $b)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT:  (drop
+  ;; RTRIP-NEXT:   (global.get $c)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT:  (drop
+  ;; RTRIP-NEXT:   (global.get $c)
+  ;; RTRIP-NEXT:  )
+  ;; RTRIP-NEXT: )
+  (func $uses
+    (drop
+      (global.get $b)
+    )
+    (drop
+      (global.get $c)
+    )
+    (drop
+      (global.get $c)
+    )
+  )
+)
+
+;; $b has more uses, but $a is an import and must remain first.
+(module
+  ;; CHECK:      (import "a" "b" (global $a i32))
+  ;; RTRIP:      (import "a" "b" (global $a i32))
+  (import "a" "b" (global $a i32))
```

Comment:
I don't see where in the code this is enforced. Is it first here because of the tie breaking with the names?

Reply:
There is no need to handle this in the pass, as both text and binary writing first emit all imports and then all non-imports. So this part of the test really just tests other parts of the codebase, I guess.

Reply:
Hmm, I thought we had an invariant that imported items came first in the IR as well. Maybe I'm wrong about that, though. Looking at `ReorderFunctions.cpp`, it doesn't look like it tries to keep imports first.

(I also noticed that `ReorderFunctions` does not take `ref.func` into account!)

Reply:
It doesn't seem like we do have that invariant. Though in practice that might be the case.

Good catch about ReorderFunctions and `ref.func` - good thing we don't run that pass yet :smile: